lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2.rb

```diff
@@ -0,0 +1,318 @@
+  TASK_TRIGGER_IDLE                  = 6
+  TASK_TRIGGER_REGISTRATION          = 7
+  TASK_TRIGGER_BOOT                  = 8
+  TASK_TRIGGER_LOGON                 = 9
+  TASK_TRIGGER_SESSION_STATE_CHANGE  = 11
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa382538%28v=vs.85%29.aspx
+  TASK_VALIDATE_ONLY                 = 0x1
+  TASK_CREATE                        = 0x2
+  TASK_UPDATE                        = 0x4
+  TASK_CREATE_OR_UPDATE              = 0x6
+  TASK_DISABLE                       = 0x8
+  TASK_DONT_ADD_PRINCIPAL_ACE        = 0x10
+  TASK_IGNORE_REGISTRATION_TRIGGERS  = 0x20
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383566(v=vs.85).aspx
+  TASK_LOGON_NONE                           = 0
+  TASK_LOGON_PASSWORD                       = 1
+  TASK_LOGON_S4U                            = 2
+  TASK_LOGON_INTERACTIVE_TOKEN              = 3
+  TASK_LOGON_GROUP                          = 4
+  TASK_LOGON_SERVICE_ACCOUNT                = 5
+  TASK_LOGON_INTERACTIVE_TOKEN_OR_PASSWORD  = 6
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa380747(v=vs.85).aspx
+  TASK_RUNLEVEL_LUA     = 0
+  TASK_RUNLEVEL_HIGHEST = 1
+
+  def self.folder_path_from_task_path(task_path)
+    path = task_path.rpartition('\\')[0]
+
+    path.empty? ? ROOT_FOLDER : path
+  end
+
+  def self.task_name_from_task_path(task_path)
+    task_path.rpartition('\\')[2]
+  end
+
+  # Returns an array of scheduled task names.
+  # By default EVERYTHING is enumerated
+  # option hash
+  #    include_child_folders: recurses into child folders for tasks. Default true
+  #    include_compatibility: Only include tasks which have any of the specified compatibility levels. Default empty array (everything is permitted)
+  #
+  def self.enum_task_names(folder_path = ROOT_FOLDER, options = {})
+    raise TypeError unless folder_path.is_a?(String)
+
+    options[:include_child_folders] = true if options[:include_child_folders].nil?
+    options[:include_compatibility] = [] if options[:include_compatibility].nil?
+
+    array = []
+
+    task_folder = task_service.GetFolder(folder_path)
+    task_folder.GetTasks(TASK_ENUM_HIDDEN).each do |task|
+      included = true
+      included = included && options[:include_compatibility].include?(task.Definition.Settings.Compatibility) unless options[:include_compatibility].empty?
+
+      array << task.Path if included
+    end
+    return array unless options[:include_child_folders]
+
+    task_folder.GetFolders(0).each do |child_folder|
+      array += enum_task_names(child_folder.Path, options)
+    end
+
+    array
+  end
+
+  def self.task(task_path)
+    raise TypeError unless task_path.is_a?(String)
+
+    task_folder = task_service.GetFolder(folder_path_from_task_path(task_path))
+
+    task_object = task_folder.GetTask(task_name_from_task_path(task_path))
+
+    task_object
+  end
+
+  def self.definition(task_object = nil)
+    definition = task_service.NewTask(0)
```

Comment:
Another weird reserved flag?

Reply:
One more comment on this method... I see it's basically designed to give you either a new definition or to clone an existing definition if you pass a task object.  Wondering if it might be better to separate the two into something like `empty_definition` and `task_definition(task)` or something like that?

I understand what you were going for... could probably go either way.

Is only existing usage in specs?

Reply:
Whoops. Agreed.  I broke the (S)olid :-) .  Changed to `new_task_definition` and `task_definition(task)`